### DISCUSSION TITLE
Updating .zimrc

### DISCUSSION THREAD

**Discussion by YonatanAhituv** [zimfw/zimfw#269] [Ideas]
Currently, in Zim, the `zimrc` is never updated.
There are many methods,
* Manual, Constantly Updated Method,
something like:
```
mv ${HOME}/.zimrc ~/.zimrc.old.$(date +%Y%m%d%H%M%S)
cat << EOF
foo=$foo
bar=()
EOF > ${HOME}/.zimrc
```
* Some sort of automatic method

* Etc.
-----

**Thoughts?**

**Comment by YonatanAhituv** [comment]
@ericbn @Eriner Thoughts?

**Comment by Plloi** [comment]
zimrc is a user config file, as such it should be updated by the user as needed/wanted.

**Comment by YonatanAhituv** [comment]
@Plloi What about commits like acdb1d4b and 52604eb2? It **is** a user configuration file but new options are added in the template and we need to reflect that in the users zimrc.

**Comment by Plloi** [comment]
No, we don't. without an option defined there should should be default behavior.
acdb1d4 will default to emacs
52604eb ~~Just alters SPROMT's default value to add color~~  Technically does nothing since commented out
not adding these to the users zimrc is fine as nothing is broken

**Comment by ericbn** [comment]
The only good way to handle changes in the zimrc template that comes to my mind is having Release Notes / a Changelog so users can check for any recent improvements there...

**Comment by PatTheMav** [comment]
What @ericbn said - `zmanage update` could show a `UPDATING` file from the git root whenever an update ran and depending on how crazy you want to go with coloring/unicode emojis breaking changes and new features could be highlighted there.

**Comment by Eriner** [comment]
We could use git to generate diffs.

**Comment by Eriner** [comment]
And just diff against the templates folder that is pulled from origin/master in with `zmanage update`. So when `zmanage update` is run, if there are differences prompt to show diffs. @ericbn

**Comment by YonatanAhituv** [comment]
@Eriner @ericbn 
Here's a starting point:
```zsh
#
# zim_update - update the zim repository
#

# Declare local vars
local diffhome
local index
local diff
local currentdir

# Get user's currentdir when running
currentdir=$(pwd)
# Make sure to clean up if quit
trap 'cd $currentdir; unset diffhome index diff currentdir; return' INT

cd ${ZIM_HOME}

# this is the cleanest way I know how to update a repository
git remote update -p
git merge --ff-only @\{u\}
# and update the submodules
git submodule update --init --recursive

# Go to templates
cd ${ZIM_HOME}/templates

# Generate list of template files that are at home
diffhome=()

for file in *; do
   diffhome+="${HOME}/.$file"
done

# Keep trap of index for diffhome
index=1

# Create diff var
diff=""

# Loop over files in templates and generate diff accordingly
for file in *; do
    echo "Diffing: $file with $diffhome[$index]..."
    =git --no-pager diff $diffhome[$index] $file && echo -ne "\n"
    index=$((index+1))
done


# Go back to the directory the user started in
cd $currentdir

# Unset local vars
unset difftemp diffhome index diff currentdir
```

**Comment by Plloi** [comment]
@AtomicCoding I feel like there would be less noise in the output if you diff the templates against their previous HEAD instead of the current file files in home. 
For exapmple if a user edits their zmodules array your version will flag it at every update, but if you diff the outgoing vs incoming you that part of the file is only flag when changes are made in the repo (new module, reorder for dependencies, etc...).
Also, my zshrc contains more than just the zim loading code, I don't need to see that as removed from the template every time i update zim.

**Comment by Plloi** [comment]
Gist of tentative version: [Link](https://gist.github.com/Plloi/0bb9218313aded516f9aa4f9703db51c)

**Comment by ericbn** [comment]
Do you feel this should be part of the `zimfw update` or `zimfw upgrade` action? Or done in some other way? Pay attention to the fact that templates are now in https://github.com/zimfw/install/tree/master/src/templates, not in the zimfw repo anymore.

**Comment by ericbn** [comment]
This is how you can manually check the diff between your dotfiles and the latest templates (replace vimdiff by your preferred diff tool):

```
for ztemplate (zimrc zlogin zshenv zshrc) vimdiff ~/.${ztemplate} https://raw.githubusercontent.com/zimfw/install/master/src/templates/${ztemplate}
```

**Comment by ericbn** [comment]
I'm announcing important updates, which includes updates to the dotfile templates, in this discussion topic: https://github.com/zimfw/zimfw/discussions/426. You can subscribe to it to receive notifications.
